pallets/asset-manager/src/lib.rs

```diff
@@ -37,79 +37,39 @@ mod tests;
+		types::AssetId,
 	};
+	use sp_runtime::{traits::AccountIdConversion, ArithmeticError};
 
 	#[pallet::pallet]
 	#[pallet::generate_store(pub(super) trait Store)]
 	#[pallet::without_storage_info]
 	pub struct Pallet<T>(_);
 
-	/// The AssetManagers's pallet id
-	pub const PALLET_ID: PalletId = PalletId(*b"asstmngr");
-
-	/// The registrar trait: defines the interface of creating an asset in the asset implementation layer.
-	/// We may revisit this interface design (e.g. add change asset interface). However, change StorageMetadata
-	/// should be rare.
-	pub trait AssetRegistrar<T: Config> {
-		/// Create an new asset.
-		///
-		/// * `asset_id`: the asset id to be created
-		/// * `min_balance`: the minimum balance to hold this asset
-		/// * `metadata`: the metadata that the implementation layer stores
-		/// * `is_sufficient`: whether this asset can be used as reserve asset,
-		/// 	to the first approximation. More specifically, Whether a non-zero balance of this asset is deposit of sufficient
-		/// 	value to account for the state bloat associated with its balance storage. If set to
-		/// 	`true`, then non-zero balances may be stored without a `consumer` reference (and thus
-		/// 	an ED in the Balances pallet or whatever else is used to control user-account state
-		/// 	growth).
-		fn create_asset(
-			asset_id: T::AssetId,
-			min_balance: T::Balance,
-			metadata: T::StorageMetadata,
-			is_sufficient: bool,
-		) -> DispatchResult;
-
-		/// Update asset metadata by `AssetId`.
-		///
-		/// * `asset_id`: the asset id to be created.
-		/// * `metadata`: the metadata that the implementation layer stores.
-		fn update_asset_metadata(
-			asset_id: T::AssetId,
-			metadata: T::StorageMetadata,
-		) -> DispatchResult;
-	}
-
-	/// The AssetMetadata trait:
-	pub trait AssetMetadata<T: Config> {
-		/// Returns the minimum balance to hold this asset
-		fn min_balance(&self) -> T::Balance;
-
-		/// Returns a boolean value indicating whether this asset needs an existential deposit
-		fn is_sufficient(&self) -> bool;
-	}
-
 	/// Convert AssetId and AssetLocation
-	impl<T: Config> AssetIdLocationGetter<T::AssetId, T::AssetLocation> for Pallet<T> {
-		fn get_asset_id(loc: &T::AssetLocation) -> Option<T::AssetId> {
+	impl<T: Config> AssetIdLocationGetter<<T::AssetConfig as AssetConfig>::AssetLocation>
+		for Pallet<T>
+	{
+		fn get_asset_id(loc: &<T::AssetConfig as AssetConfig>::AssetLocation) -> Option<AssetId> {
 			LocationAssetId::<T>::get(loc)
 		}
 
-		fn get_asset_location(id: T::AssetId) -> Option<T::AssetLocation> {
+		fn get_asset_location(
+			id: AssetId,
+		) -> Option<<T::AssetConfig as AssetConfig>::AssetLocation> {
 			AssetIdLocation::<T>::get(id)
 		}
 	}
 
 	/// Get unit per second from `AssetId`
-	impl<T: Config> UnitsToWeightRatio<T::AssetId> for Pallet<T> {
-		fn get_units_per_second(id: T::AssetId) -> Option<u128> {
+	impl<T: Config> UnitsToWeightRatio for Pallet<T> {
+		fn get_units_per_second(id: AssetId) -> Option<u128> {
```

Comment:
I think we are going to make generic code as possible, but there are two exceptions, `Balance` and `AssetId`, these two types are hard-wired into ZKP and we don't want to change them in the foreseeable future. So making them concrete could simplify the code.

Reply:
Also, the code need not be generic at the interface level since we aren't creating two instances of pallet-manta-pay with different `AssetId` types. Instead if we want to change the _unique_ instance of pallet-manta-pay, we just change the `AssetId` type.